From 5186db271b48606c4a5202384a398c39c5e6e377 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Fri, 30 Nov 2018 16:15:40 +0900
Subject: [PATCH 337/909] drm/bridge: adv7511: Add function of suspend and
 resume

commit f8bc2cd28575d2c483d15116d41810b8b9410026 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

The functions of suspend and resume are supported by this patch.
However, if the hotplug process runs at resume, the device driver that
will be raised by the pm_runtime process will return EACCES at executing
pm_runtime_get, so clear hotplug interrupt once at resume so that hotplug
event will not be executed.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/adv7511/adv7511_drv.c | 28 ++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c b/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
index dd3ff2f2cdce..dbc464259a5f 100644
--- a/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
+++ b/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
@@ -1262,6 +1262,33 @@ static int adv7511_remove(struct i2c_client *i2c)
 	return 0;
 }
 
+#ifdef CONFIG_PM_SLEEP
+static int adv7511_pm_suspend(struct device *dev)
+{
+	struct i2c_client *i2c = to_i2c_client(dev);
+	struct adv7511 *adv7511 = i2c_get_clientdata(i2c);
+
+	adv7511_power_off(adv7511);
+
+	return 0;
+}
+
+static int adv7511_pm_resume(struct device *dev)
+{
+	struct i2c_client *i2c = to_i2c_client(dev);
+	struct adv7511 *adv7511 = i2c_get_clientdata(i2c);
+
+	adv7511_detect(adv7511, &adv7511->connector);
+	adv7511_power_on(adv7511);
+
+	return 0;
+}
+#endif
+
+static const struct dev_pm_ops adv7511_pm_ops = {
+	SET_LATE_SYSTEM_SLEEP_PM_OPS(adv7511_pm_suspend, adv7511_pm_resume)
+};
+
 static const struct i2c_device_id adv7511_i2c_ids[] = {
 	{ "adv7511", ADV7511 },
 	{ "adv7511w", ADV7511 },
@@ -1292,6 +1319,7 @@ static struct i2c_driver adv7511_driver = {
 	.driver = {
 		.name = "adv7511",
 		.of_match_table = adv7511_of_ids,
+		.pm = &adv7511_pm_ops,
 	},
 	.id_table = adv7511_i2c_ids,
 	.probe = adv7511_probe,
-- 
2.17.1

